lib/mutant/expression/parser.rb

```diff
@@ -0,0 +1,74 @@
+module Mutant
+  class Expression
+    class Parser
+      include Concord.new(:types)
+
+      class ParserError < RuntimeError
+        include AbstractType
+      end
+
+      # Error raised on invalid expressions
+      class InvalidExpressionError < ParserError; end
+
+      # Error raised on ambiguous expressions
+      class AmbiguousExpressionError < ParserError; end
+
+      # Parse input into expression or raise
+      #
+      # @param [String] syntax
+      #
+      # @return [Expression]
+      #   if expression is valid
+      #
+      # @raise [ParserError]
+      #   otherwise
+      #
+      # @api private
+      #
+      def call(input)
+        try_parse(input) or fail InvalidExpressionError, "Expression: #{input.inspect} is not valid"
+      end
+
+      # Try to parse input into expression
+      #
+      # @param [String] input
+      #
+      # @return [Expression]
+      #   if expression is valid
+      #
+      # @return [nil]
+      #   otherwise
+      #
+      # @api private
+      #
+      def try_parse(input)
+        expressions = expressions(input)
+        case expressions.length
+        when 0, 1
+          expressions.first
+        else
+          fail AmbiguousExpressionError, "Ambiguous expression: #{input.inspect}"
+        end
+      end
+
+    private
+
+      # Return expressions for input
+      #
+      # @param [String] input
+      #
+      # @return [Array<Expression>]
+      #   if expressions can be parsed from input
+      #
+      # @api private
+      #
+      def expressions(input)
+        types.each_with_object([]) do |type, aggregate|
+          expression = type.try_parse(input)
+          aggregate << expression if expression
```

Comment:
This is totally OT, but I feel like there should be an enumerable method that works like `Enumerable#select`, but returns the "truthy" return values from the block like `#map`. This would eliminate the need for using `#each_with_object` like this.